### ISSUE TITLE
[clang] Defaulted method template specialization fails to compile

### ISSUE THREAD

**Issue by nmakeenkov** [llvm/llvm-project#66772]
Tried the example on godbolt with clang 16.0.0: https://gcc.godbolt.org/z/K1dGc55Tx

I have a template class with defaulted copy assignment operator in base declaration. I want to specialise it for some template parameter and get an error `error: definition of explicitly defaulted copy assignment operator`.
But I believe it's legal to define this operator.

The code is as follows (with g++ compiles and works as expected):
```
#include <iostream>

template <typename T>
class MyClass {
public:
    MyClass() {}
    MyClass& operator=(const MyClass& other) = default;
};

// Specialization
template <>
MyClass<int>& MyClass<int>::operator=(const MyClass& other) noexcept {
    std::cout << "Specialized optor for int" << std::endl;
    return *this;
}

int main() {
    MyClass<double> obj1;
    MyClass<double> obj2;

    obj1 = obj2;

    MyClass<int> obj3;
    MyClass<int> obj4;

    obj4 = obj3;

    return 0;
}
```

**Comment by shafik** [comment]
Interestingly gcc seems to be the only one that allows this: https://gcc.godbolt.org/z/eonnh5hvn

AFAICT there is a restriction on *implicity-declared* special member functions but not *defaulted* in [temp.expl.spec p5](https://eel.is/c++draft/temp.expl.spec#5):

> ... If such an explicit specialization for the member of a class template names an implicitly-declared special member function ([special]), the program is ill-formed.

We can see this in action where everyone agrees if we comment out the defaulted operator: https://gcc.godbolt.org/z/GEGzE3MGd

I don't think that [temp.expl.spec]p5 applies here but perhaps I am misunderstanding. I can see it both ways but not totally sure.

CC @zygoloid @hubert-reinterpretcast

**Comment by cor3ntin** [comment]
@shafik I don't think it applies either. 
I think it probably make sense to want to do something other than the default for a specialization, so I assume the wording reflects the intent of core (although it might be good to confirm).
So clang is probably not conforming.

**Comment by zygoloid** [comment]
I don't think we can reasonably support code such as this. Class properties, such as whether it's trivially-copyable, shouldn't be able to change once the class is complete. But what about this:

```c++
template<typename T> struct MyClass {
  MyClass& operator=(const MyClass& other) = default;
};

static_assert(__is_trivially_copyable(MyClass<int>));

template<> MyClass<int>& MyClass<int>::operator=(const MyClass& other) noexcept {
  return *this;
}

static_assert(!__is_trivially_copyable(MyClass<int>));
```

GCC claims that `MyClass<int>` is trivially copyable after the explicit specialization is defined. (And this isn't a caching problem: GCC gives the same result if we comment out the first `static_assert`.)

Also, for a copy *constructor* that is specialized in this way, GCC does *not* call the explicit specialization to actually copy the object, and performs a bit-copy instead: https://gcc.godbolt.org/z/qTja8oP96

**Comment by shafik** [comment]
> I've mailed the core reflector suggesting we reject: https://lists.isocpp.org/core/2023/09/14866.php

Thank you, that was actually on my list of things todo ✅
